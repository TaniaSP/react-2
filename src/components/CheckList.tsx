import React, { useEffect, useMemo, useState } from 'react'
import PropTypes from 'prop-types'
import useComponentVisible from '../services/hooks'
interface Opt {
  value: string
  checked: boolean
}

export default function CheckList ({ value, onSelected }: { value: string[], onSelected: Function }): JSX.Element {
  const { ref, isComponentVisible } = useComponentVisible(true)

  const DEFAULT_GENRES = ['Adventure', 'Comedy', 'Drama', 'Science Fiction', 'War', 'Romance', 'Horror', 'Action', 'Thriller', 'Mystery', 'Fantasy', 'Animation', 'Family', 'Music', 'Crime', 'Western', 'History']

  const opts: Opt[] = DEFAULT_GENRES.map(x =>
    ({ checked: value.find(v => v === x) !== undefined, value: x })
  )
  const [isOpen, setIsOpen] = useState(false)

  const updateSelcted = (item: string): void => {
    const updated: Opt[] = opts.map((x: Opt) => ({
      ...x,
      checked: x.value === item ? !x.checked : x.checked
    }))
    onSelected(updated.filter(x => x.checked).map(x => x.value))
  }

  useEffect(() => {
    if (!isComponentVisible) {
      setIsOpen(false)
    }
  }, [isComponentVisible])

  const fullSelection = useMemo(() => value.length > 0 ? value.join(', ') : 'Select Genre', [value])

  return (
    <div className="list-droprown">
      <button type="button" onClick={() => setIsOpen(!isOpen)} className={isOpen ? 'open' : ''} title={fullSelection}>{fullSelection}</button>
      {isOpen && <ul ref={ref} className="checkbox-list">
        {opts.map((x: Opt, i: number) => <li key={x.value}> <label htmlFor={`check${i}`}> <input id={`check${i}`} type="checkbox" checked={x.checked} onChange={() => updateSelcted(x.value)} />{x.value}</label></li>)}
      </ul>}
    </div>
  )
}

CheckList.defaultProps = {
  value: []
}
CheckList.propTypes = {
  value: PropTypes.array
}
