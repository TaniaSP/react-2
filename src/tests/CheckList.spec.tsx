import React from 'react'
import { render, screen, act } from '@testing-library/react'
import '@testing-library/jest-dom'
import CheckList from '../components/CheckList'
import { BrowserRouter as Router } from 'react-router-dom'
beforeEach(() => {
  jest.mock('../services/hooks', () => {
    return jest.fn().mockImplementation(() => ({
      ref: { current: [] }, isComponentVisible: true, setIsComponentVisible: () => { }
    }))
  })
})

test('should render Checklist component with query prop', async () => {
  // ARRANGE
  const value: string[] = ['Adventure', 'Comedy']
  const method: Function = () => { }
  const { asFragment } = render(<Router><CheckList value={value} onSelected={method} /></Router>)
  const button = await screen.findByTitle('Adventure, Comedy')
  act(() => button.click())
  const adventureItem = await screen.findByLabelText('Adventure')
  act(() => adventureItem.click())
  // ASSERT
  expect(asFragment).toMatchSnapshot()
  expect(screen.getByTitle(value.join(', '))).toBeDefined()
})
