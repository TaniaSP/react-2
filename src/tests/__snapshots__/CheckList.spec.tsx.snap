// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Checklist component with query prop 1`] = `[Function]`;
